gax/src/streamingRetryRequest.ts

```diff
@@ -0,0 +1,204 @@
+// Copyright 2023 Google LLC
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     https://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+const {PassThrough} = require('stream');
+
+const DEFAULTS = {
+  /*
+    Max # of retries
+  */
+  maxRetries: 2,
+
+  /*
+    The maximum time to delay in seconds. If retryDelayMultiplier results in a
+    delay greater than maxRetryDelay, retries should delay by maxRetryDelay
+    seconds instead.
+  */
+  maxRetryDelayMillis: 64000,
+
+  /*
+    The multiplier by which to increase the delay time between the completion of
+    failed requests, and the initiation of the subsequent retrying request.
+  */
+  retryDelayMultiplier: 2,
+
+  /*
+    The length of time to keep retrying in seconds. The last sleep period will
+    be shortened as necessary, so that the last retry runs at deadline (and not
+    considerably beyond it).  The total time starting from when the initial
+    request is sent, after which an error will be returned, regardless of the
+    retrying attempts made meanwhile.
+  */
+  totalTimeoutMillis: 600000,
+
+  /*
+    The initial delay time, in milliseconds, between the completion of the first
+    failed request and the initiation of the first retrying request.
+  */
+  initialRetryDelayMillis: 60,
+
+  /*
+    Initial timeout parameter to the request.
+  */
+  initialRpcTimeoutMillis: 60,
+
+  /*
+    Maximum timeout in milliseconds for a request.
+    When this value is reached, rpcTimeoutMulitplier will no
+    longer be used to increase the timeout.
+  */
+  maxRpcTimeoutMillis: 60,
+
+  /*
+   Multiplier by which to increase timeout parameter in
+   between failed requests.
+  */
+  rpcTimeoutMultiplier: 2,
+
+  /*
+    The number of retries that have occured.
+  */
+  retries: 0,
+
+  retryCodesOrShouldRetryFn:
+    [14] ||
+    function (response: any) {
+      return undefined;
+    },
+
+  getResumptionRequestFn: function (response: any) {
+    return undefined;
+  },
+};
+
+export function streamingRetryRequest(
+  requestOpts: any = null,
```

Comment:
If you don't care about the content, any options object would be `{} | null`, and the callback will be `Function | null`, or something like that. Preferably not an `any`.

Reply:
We actually only ever pass null for requestOpts and for the callback - I'm doing a refactor to remove what we don't even use, and I'll fix the any's for what's left